### DISCUSSION TITLE
Server crash after big file upload

### DISCUSSION THREAD

**Discussion by aradox** [pocketbase/pocketbase#835] [Q&A]
Hey,
When i try to upload a big file (2gb) the server crashes, and i get a 502 error. I already changed the max_file_size in Nginx to support big files but i believe the error is in pocketbase. Does anybody have a solution for this?

![image](https://user-images.githubusercontent.com/4798993/197713832-f240f4f9-c86c-4d08-bfd0-e49a34685580.png)

i set the max upload size to 3GB
![image](https://user-images.githubusercontent.com/4798993/197714270-359cf14f-dad0-461a-9bdd-30c005753375.png)

**Comment by ganigeorgiev** [comment]
PocketBase doesn't use 502 as response code (at least not directly) so my guess is that this is something related to the nginx configuration (or maybe client-side timeout during upload?).

Could you try reuploading the file with a local PocketBase instance without nginx?

To check for PocketBase errors you could start the executable with `--debug` and/or preview the error request logs from the "Admin UI > Logs".

**Reply by ganigeorgiev** [reply]
Additionally, you may also want to experiment with the various available nginx configurations like [`proxy_max_temp_file_size`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_max_temp_file_size), [`proxy_read_timeout`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout), [`proxy_send_timeout`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_send_timeout), [`client_body_timeout`](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_body_timeout), [`send_timeout`](https://nginx.org/en/docs/http/ngx_http_core_module.html#send_timeout), etc. (I don't think `fastcgi_*` directives has any effect when used with PocketBase).

**Reply by aradox** [reply]
I did try it on my local machine and i can just upload the same file. After looking into the server logs i found this error _Out of memory: Killed process 684 (pocketbase) total-vm:1096688kB._ Then I resized my droplet to 2GB memory and it still doesnt work.

![image](https://user-images.githubusercontent.com/4798993/197719788-d3108c6b-eefb-4d0a-9707-9268bad337ae.png)

**Reply by ganigeorgiev** [reply]
In order to do the necessary file validations, PocketBase reads the file content in memory and this is definitely not optimal for large file uploads as in your case. Unfortunately, other than increasing the ram, there is no easy solution for that at the moment.

I've created an issue to improve large files upload/serve - https://github.com/pocketbase/pocketbase/issues/836, but it will be handled after the v0.8 prerelease.

**Reply by ganigeorgiev** [reply]
@aradox Just to let you know that I've managed to implement a fix locally and it will be available with the v0.8.0 pre-release (most likely this weekend). It performs slightly slower since now we read the file in chunks from the disk, but it consumes a lot less memory (~50MB with 1GB local file test) and it is certainly better than throwing OOM errors.

**Reply by ganigeorgiev** [reply]
@aradox Just to let you know that the fix is available as part of the [v0.8.0-rc1 pre-release](https://github.com/pocketbase/pocketbase/releases/tag/v0.8.0-rc1).
